### PR TITLE
Clarity Value Serialization and Name String Guarding

### PR DESCRIPTION
**Pull request by kantai**
This PR implements a standard serialization and deserialization scheme for Clarity `Value`s -- they are serialized to a subset of legal JSON strings. The deserializer uses Serde to parse the JSON, so the deserializer is _more_ permissive than the serializer.

* Changes the deserialization interface of Clarity values to have two options:
   * With an expected type: this is how the Clarity VM loads data out of the database: it has an expectation of the type (because of the type declarations of data maps, data vars, etc.)
   * WIthout an expected type: this is how the wire format deserializer should call it.
* Refactor of the `types.rs` file into a folder module -- it had grown pretty unwieldy.
* Some refactoring of the TypeSignature and Value code to eliminate repetition, dead code, and some increased test coverage.

Other objects will _also_ need special serializers (anything that goes into the MARF should have a specified serialization), but that's going to be left for later work, since the wire format won't block on those.

This PR also adds guarding to the representation of names in Clarity (implementation of #1100) -- Contract names and general names (variables, functions, tuple fields) all now use one of two special types:

```
ClarityName
ContractName
```

Which are guarded wrappers around `String` types. These guards ensure that any such names are valid Clarity names (a subset of ASCII).

### CODE REVIEW

File: src/vm/types/mod.rs [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+impl PrincipalData {
+    pub fn parse_qualified_contract_principal(literal: &str) -> Result<PrincipalData> {
+        let split: Vec<_> = literal.splitn(2, ".").collect();
+        if split.len() != 2 {
+            return Err(RuntimeErrorType::ParseError(
+                "Invalid principal literal: expected a `.` in a qualified contract name".to_string()).into());
+        }
+        let sender = Self::parse_standard_principal(split[0])?;
+        let name = split[1].to_string();
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: Can we be sure that `name` is an ASCII string?

File: src/vm/types/serialization.rs [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+        #[serde(rename="type")]
+        type_n: String,
+        issuer: String,
+        name: String
+    }
+}
+
+fn to_hex(val: &i128) -> String {
+    if *val >= 0 {
+        format!("{:x}", val)
```
Comment (jcnelson) [inline src/vm/types/serialization.rs]: Maybe this is a paranoid delusion on my part, but how stable can we expect formatting rules to be in practice?  I'm mainly concerned about whether or not the user's locale settings can affect the formatter's behavior (e.g. by making characters upper-case, by tacking on one or more leading 0's, by flipping the endianness, and so on).

File: src/vm/types/serialization.rs [hunk 3]
```diff
@@ -0,0 +1,862 @@
...
+                let type_name = if response.committed {
+                    TYPE_RESP_OK
+                } else {
+                    TYPE_RESP_ERR
+                };
+                let value = response.data.serialize();
+                json_recursive_object(type_name, &value)
+            },
+            Optional(OptionalData{ data: None }) => {
+                format!(r#"{{ "type": "{}" }}"#, TYPE_OPT_NONE)
```
Comment (jcnelson) [inline src/vm/types/serialization.rs]: I ran into this problem when writing the original SIP 005 spec:  do we need to encode here what the type of the `Some` variant is?  There is ambiguity by just having `{"type": "none"}` -- `none` could be type `Option<int>` or `Option<bool>`, for example.

Comment (jcnelson) [reply src/vm/types/serialization.rs]: Maybe this isn't important because `none` admits any type?

File: src/vm/types/serialization.rs [hunk 4]
```diff
@@ -0,0 +1,862 @@
...
+                                format!(r#""{}": {}"#, key, value.serialize()))
+                    .collect();
+                let entries_str = entries.join(", ");
+                format!(
+                    r#"{{ "type": "{}", "entries": {{ {} }} }}"#,
+                    TYPE_TUPLE, entries_str)
+            }
+        }
+    }
+}
```
Comment (jcnelson) [inline src/vm/types/serialization.rs]: I think this is the case, but I just want to be sure it's been considered:  there's _no possible way_ for a value to serialize to anything that has JSON special characters, right?  i.e. no `,`, `{`, `[`, `}`, `]`, or `"`?

File: src/vm/types/mod.rs (hunk 1 above)
Comment (kantai) [reply src/vm/types/mod.rs]: Not as written -- but that's going to be addressed in #1078 (is my understanding), at which point we should have a simple guarded string type.

File: src/vm/types/serialization.rs (hunk 3 above)
Comment (kantai) [reply src/vm/types/serialization.rs]: For the Option Values themselves, the type doesn't need to be encoded (it's not stored in memory, for example, during evaluation) -- only lists (because they have a maximum length, separate from their actual length) and tuples (which may not actually need their type maps) store their types.

File: src/vm/types/serialization.rs (hunk 4 above)
Comment (kantai) [reply src/vm/types/serialization.rs]: That's correct -- the only "strings" that are serialized are tuple names, and currently, the parser will not allow those to contain any of those values. However, like the other Strings in use through the VM, we should have guarded String types for these values.

Comment (jcnelson) [reply src/vm/types/serialization.rs]: Sounds good.  As long as #1078 will introduce a guarded string type, then I think we're good here.

File: src/vm/types/serialization.rs (hunk 2 above)
Comment (kantai) [reply src/vm/types/serialization.rs]: I share this fear, so it's at least a shared delusion.

Comment (kantai) [reply src/vm/types/serialization.rs]: I spent a while poking at the LowerHex docs, but produced no satisfactory specification, so I'll just update this to do the hexing manually.

Comment (jcnelson) [reply src/vm/types/serialization.rs]: Thank you for feeding my paranoia and general mistrust of GNU libc locale settings :ghost:

File: src/vm/types/serialization.rs [hunk 5]
```diff
@@ -0,0 +1,881 @@
...
+
+impl Value {
+    pub fn deserialize(json: &str, expected: &TypeSignature) -> Self {
+        Value::try_deserialize(json, expected)
+            .expect("ERROR: Failed to parse Clarity JSON")
+    }
+}
+
+
+#[cfg(test)]
```
Comment (jcnelson) [inline src/vm/types/serialization.rs]: I'm not sure how reasonable or difficult this is, so feel free to dismiss.

Would it be desirable to add an "exhaustive" test that tries to serialize/deserialize all possible type combinations up to a certain (small) recursion depth?  Might be the job of a fuzzer to do this; just something to think about.

Comment (kantai) [reply src/vm/types/serialization.rs]: Yeah -- the Value enum is fairly amenable to such a test.

File: src/vm/representations.rs [hunk 6]
```diff
@@ -32,6 +29,55 @@ pub struct SymbolicExpression {
...
+        #[cfg(test)]
+        impl From<&'_ str> for $Name {
+            fn from(value: &str) -> Self {
+                Self::try_from(value.to_string()).unwrap()
+            }
+        }
+    }
+}
+
+guarded_string!(ClarityName, "ClarityName", Regex::new("^([a-zA-Z]|[-!?+<>=/*])([a-zA-Z0-9]|[-_!?+<>=/*])*$"));
```
Comment (jcnelson) [inline src/vm/representations.rs]: Can ClarityNames start with punctuation?

File: src/vm/types/mod.rs [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+// a standard principal is a version byte + hash160 (20 bytes)
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct StandardPrincipalData(pub u8, pub [u8; 20]);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum PrincipalData {
+    StandardPrincipal(StandardPrincipalData),
+    ContractPrincipal(ContractName),
+    QualifiedContractPrincipal { sender: StandardPrincipalData,
+                                 name: ContractName },
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: Not sure if this deserves a top-level issue (it's already present in the versioning issue), but a qualified contract principal will also take a nonce.  This would be done in a separate PR though; I'm only raising it to ask how you want the issue recorded.

File: src/vm/types/mod.rs [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+    Tuple(TupleData),
+    Optional(OptionalData),
+    Response(ResponseData)
+}
+
+define_named_enum!(BlockInfoProperty {
+    Time("time"),
+    VrfSeed("vrf-seed"),
+    HeaderHash("header-hash"),
+    BurnchainHeaderHash("burnchain-header-hash"),
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: Don't we also need the block height and the parent header hash?

File: src/vm/types/mod.rs [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+
+        Ok(Value::List(ListData { data: list_data, type_signature: expected_type }))
+    }
+
+    pub fn list_from(list_data: Vec<Value>) -> Result<Value> {
+        let type_sig = TypeSignature::construct_parent_list_type(&list_data)?;
+        // Aaron: at this point, we've _already_ allocated memory for this type.
+        //     (e.g., from a (map...) call, or a (list...) call.
+        //     this is a problem _if_ the static analyzer cannot already prevent
+        //     this case. This applies to all the constructor size checks.
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: Would this mean that a high-dimensional list with one large dimension value could either (1) trigger OOM or (2) bill the contract creator with a high tx fee as if the list were a high-dimensional "cube", like before?  We probably want to avoid both (1) and (2) if at all possible.

Comment (kantai) [reply src/vm/types/mod.rs]: Yep -- the type checker will need to save us in this case, which it will be able to do (once we merge https://github.com/blockstack/blockstack-core/tree/feature/value-size-bounding), since two properties are true:

1. TypeSignatures will Error rather than construct a too-large-value.
2. Every node in the AST will be associated with a TypeSignature following a successful type_checker pass.

File: src/vm/types/mod.rs (hunk 8 above)
Comment (kantai) [reply src/vm/types/mod.rs]: `block-height` is a native global, rather than an argument to the `block-info` function. It's defined in the `variables.rs` file.

File: src/vm/types/mod.rs (hunk 7 above)
Comment (kantai) [reply src/vm/types/mod.rs]: Yeah -- I think we just log that in the versioning issue, and ensure that the PR implementing that issue makes that change.

File: src/vm/representations.rs (hunk 6 above)
Comment (kantai) [reply src/vm/representations.rs]: Currently, yes -- things like the arithmetic operators are legal names (`+`, `-`, etc.).

However, we could change this so that the arithmetic operators are legal names, but nothing else that starts with punctuation is:

```
(^([a-zA-Z])([a-zA-Z0-9]|[-_!?+<>=/*])*)|^[-+=/*]$|^[<>][=]?$
```

Which I think is a good idea -- will change that.

File: src/vm/types/mod.rs [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+        if type_sig.size()? > MAX_VALUE_SIZE {
+            return Err(RuntimeErrorType::ValueTooLarge.into())
+        }
+        Ok(Value::List(ListData { data: list_data, type_signature: type_sig }))
+    }
+
+    pub fn buff_from(buff_data: Vec<u8>) -> Result<Value> {
+        if buff_data.len() > u32::max_value() as usize {
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: Is this check necessary, given that `MAX_VALUE_SIZE` is smaller than 4 billion bytes?  Is there a material difference between `BufferTooLarge` and `ValueTooLarge`?

File: src/vm/types/mod.rs [hunk 11]
```diff
@@ -0,0 +1,450 @@
...
+            Err(RuntimeErrorType::BufferTooLarge.into())
+        } else if buff_data.len() as i128 > MAX_VALUE_SIZE {
+            Err(RuntimeErrorType::ValueTooLarge.into())
+        } else {
+            Ok(Value::Buffer(BuffData { data: buff_data }))
+        }
+    }
+
+    pub fn size(&self) -> Result<i128> {
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: How can this method fail?  What does it mean for it to fail?

File: src/vm/types/mod.rs [hunk 12]
```diff
@@ -0,0 +1,450 @@
...
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    #[test]
+    fn test_constructors() {
+        assert_eq!(
+            Value::list_with_type(
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: Can we add a test with higher list dimensions?

File: src/vm/types/serialization.rs [hunk 13]
```diff
@@ -0,0 +1,884 @@
...
+
+                let (list_type, entry_type) = match expected_type {
+                    None => (None, None),
+                    Some(TypeSignature::List(list_type)) => (Some(list_type), Some(list_type.get_list_item_type())),
+                    Some(x) => return Err(InterpreterError::DeserializeExpected(x.clone()).into())
+                };
+
+                let items: InterpreterResult<_> = entries
+                    .drain(..)
+                    .map(|value| Value::try_deserialize_parsed(value, entry_type.as_ref()))
```
Comment (jcnelson) [inline src/vm/types/serialization.rs]: Should this method bound its recursion, i.e. as a safety/sanity check?

File: src/vm/types/mod.rs (hunk 10 above)
Comment (kantai) [reply src/vm/types/mod.rs]: Nope -- though in some indeterminate future when it MAX_VALUE_SIZE may be way too large, this check will be necessary. Also, no, there's no reason for both BufferTooLarge and ValueTooLarge to exist. BufferTooLarge is eliminated in the value-bounding branch. (For what it's worth, this code isn't really new in this PR -- it's just appearing as new because of breakings types.rs into mod.rs, signatures.rs, and serialization.rs)

File: src/vm/types/signatures.rs [hunk 14]
```diff
@@ -132,367 +72,67 @@ impl FunctionType {
...
-            },
-            PrincipalData::ContractPrincipal(contract_name) => {
-                write!(f, "'CT{}", contract_name)
-            },
-            PrincipalData::QualifiedContractPrincipal { sender, name } => {
-                let c32_str = c32::c32_address(sender.0, &sender.1[..])
-                    .unwrap_or_else(|_| "INVALID_C32_ADD".to_string());
-                write!(f, "'CT{}.{}", c32_str, name)
-            }
+    pub fn size(&self) -> Result<i128> {
```
Comment (jcnelson) [inline src/vm/types/signatures.rs]: Do you think we could structure lists so that the size calculation simply can't overflow?  We'd need to make it so you can't construct a list that has the capacity for so many items that an overflow in this method is possible (which is probably desirable just for resource utilization purposes).

File: src/vm/types/mod.rs (hunk 11 above)
Comment (kantai) [reply src/vm/types/mod.rs]: Similar to the u32 max value check -- type sizes should _never_ result in arithmetic overflows, but that's a property of the MAX_VALUE_SIZE check, not the Rust types. Furthermore, in the value bounding branch, `.size()` is the only method that is allowed to be invoked during TypeSignature construction (meaning its the only function which may be called on a too-large type), and is used for the actual MAX_VALUE_SIZE enforcement, meaning that some code _could_ conceivably result in an arithmetic overflow in size().

File: src/vm/types/mod.rs (hunk 12 above)
Comment (kantai) [reply src/vm/types/mod.rs]: That following test (line 412) does use higher dimensions.

File: src/vm/types/mod.rs (hunk 11 above)
Comment (jcnelson) [reply src/vm/types/mod.rs]: I feel like this is a risky approach.  I think it might be prudent to make the TypeSignature construction itself fail if the size would be too big (thereby making such type unrepresentable), and in so doing, deny the instantiation of Clarity code that could produce unrepresentable types.  What are your thoughts?

Comment (kantai) [reply src/vm/types/mod.rs]: That's exactly what it will do -- type signature constructors are all guarded. However, rather than implement size computation twice, the constructors call `size` to check whether or not the size of the type is too large -- this means that the size function _needs_ to be able to handle the case in which a type signature construction will fail. As for the size() invocation in Value, it should never fail, and this can be enforced with an `expects()`.

File: src/vm/types/serialization.rs (hunk 13 above)
Comment (kantai) [reply src/vm/types/serialization.rs]: I don't think so -- any recursion happening here has already needed to perform the same recursion during the construction of the value.

**Comment by kantai** [comment]
For the size checks, this is what is mentioned in the value bounding branch:

//   1. A TypeSignature constructor will always fail rather than construct a
//        type signature for a too large or invalid type. This is why any variable length
//        type signature has a guarded constructor.
//   2. The only methods which may be called on TypeSignatures that are too large
//        (i.e., the only function that can be called by the constructor before
//         it fails) is the `.size()` method, which may be used to check the size.

The rationale for (1) is that enforcing a size check using the rust type system would require rust to have support for dependent types. Enforcing through guarded constructors (and private instance variables) is, I think, the next best approach.

The rationale for (2) is that the size checks would otherwise need to be implemented twice: once for use in constructors, and second for use elsewhere. I can split this into a private function (fallible) and a public wrapper (infallible).

File: src/vm/types/signatures.rs (hunk 14 above)
Comment (kantai) [reply src/vm/types/signatures.rs]: Yep -- I think this will be more or less addressed in the value bounding branch. It will still perform checked arithmetic for reasons mentioned elsewhere, but in non-construction contexts, I think it can be made infallible.

File: src/vm/representations.rs (hunk 6 above)
Comment (kantai) [reply src/vm/representations.rs]: Updated in 74257cdc

**Comment by jcnelson** [comment]
Sounds good to me.  As long as this is handled in another branch, then I'm good to go on this PR.

File: src/vm/functions/database.rs [hunk 15]
```diff
@@ -43,7 +42,7 @@ pub fn special_fetch_variable(args: &[SymbolicExpression],
     let var_name = args[0].match_atom()
         .ok_or(UncheckedError::ExpectedVariableName)?;
 
-    env.global_context.database.lookup_variable(&env.contract_context.name, var_name)
+    env.global_context.database.lookup_variable(&&env.contract_context.name, var_name)
```
Comment (lgalabru) [inline src/vm/functions/database.rs]: is `&&` a sign that we're doing something wrong?

File: src/vm/parser.rs [hunk 16]
```diff
@@ -76,7 +77,7 @@ pub fn lex(input: &str) -> Result<Vec<(LexItem, u32, u32)>> {
         LexMatcher::new("'CT(?P<value>[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41}.([[:alpha:]]|[-]){5,40})", TokenType::QualifiedContractPrincipalLiteral),
         LexMatcher::new("'CT(?P<value>([[:alpha:]]|[-]){5,40})", TokenType::ContractPrincipalLiteral),
         LexMatcher::new("'(?P<value>[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41})", TokenType::PrincipalLiteral),
-        LexMatcher::new("(?P<value>([[:word:]]|[-#!?+<>=/*])+)", TokenType::Variable),
+        LexMatcher::new("(?P<value>([[:word:]]|[-!?+<>=/*])+)", TokenType::Variable),
```
Comment (lgalabru) [inline src/vm/parser.rs]: Aren't we supposed to escape `?`, `+` and  `*` ?
I did notice while working on #1078 that the `QualifiedContractPrincipalLiteral ` regex is not working as expected, because of the "." not being escaped.

File: src/vm/representations.rs [hunk 17]
```diff
@@ -32,6 +29,55 @@ pub struct SymbolicExpression {
...
+        #[cfg(test)]
+        impl From<&'_ str> for $Name {
+            fn from(value: &str) -> Self {
+                Self::try_from(value.to_string()).unwrap()
+            }
+        }
+    }
+}
+
+guarded_string!(ClarityName, "ClarityName", Regex::new("^[a-zA-Z]([a-zA-Z0-9]|[-_!?+<>=/*])*$|^[-+=/*]$|^[<>]=?$"));
```
Comment (lgalabru) [inline src/vm/representations.rs]: Same question here.

File: src/vm/parser.rs (hunk 16 above)
Comment (kantai) [reply src/vm/parser.rs]: inside of regex brackets (`[` and `]`), those characters shouldn't be escaped.

File: src/vm/functions/database.rs (hunk 15 above)
Comment (kantai) [reply src/vm/functions/database.rs]: I'm pretty sure that this can be safely changed to `&`, but I want to make sure that this wasn't a change in rustc `1.35`

File: src/vm/parser.rs (hunk 16 above)
Comment (kantai) [reply src/vm/parser.rs]: Though you're right, that outside of the brackets, those should be escaped (which is why the `.` needs to be escaped in the `QualifiedContractPrincipalLiteral`)

File: src/vm/functions/database.rs (hunk 15 above)
Comment (kantai) [reply src/vm/functions/database.rs]: Updated

**Comment by kantai** [comment]
Merging! :metal:
